### PR TITLE
chore(core): allow declaring batch upgrades

### LINKED ISSUE #10513: Allow core to declare Elgg\Upgrade\Batch upgrades
Currently it is possible to declare `Elgg\Upgrade\Batch` upgrades only through `elgg-plugin.php` file therefore preventing core from using the upgrade interface.

We could simply add a file `engine/upgrades.php` that would return array of upgrades.

### PR DESCRIPTION
**Pull request by juho-jaakkola**
Fixes #10513

### CODE REVIEW

File: engine/classes/Elgg/Upgrade/Locator.php [hunk 1]
```diff
@@ -58,9 +58,32 @@ public function run() {
...
+
+			foreach ($batches as $class) {
+				if ($this->getUpgrade($class, $plugin_id)) {
+					$pending_upgrades = true;
+				}
+			}
+		}
+
+		// TODO Where should the upgrades be declared?
+		$core_upgrades = [
```
Comment (hypeJunction) [inline engine/classes/Elgg/Upgrade/Locator.php]: `_elgg_get_upgrades()` perhaps?

Comment (jdalsem) [inline engine/classes/Elgg/Upgrade/Locator.php]: core upgrade should check/run before any plugin upgrade is checked/run...

Comment (juho-jaakkola) [inline engine/classes/Elgg/Upgrade/Locator.php]: Good point. Core upgrades are now checked first.

**Comment by jdalsem** [comment]
We could use some docs for writing core upgrades...

**Comment by mrclay** [comment]
@juho-jaakkola In what case would an upgrade want to have isRequired() return false?

**Comment by mrclay** [comment]
@juho-jaakkola all my questions: :)

- In what case would an upgrade want to have isRequired() return false?
- Why do we call countItems() on every request to the backend? What if we can't get a count of how many there are remaining to process, but we know when we're done?

**Comment by mrclay** [comment]
Replaced by #10879

**Pull request closed without merging**